cmd/pmp/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+//! Often paired with `humility map`.
+//!
+//! To better understand the memory that a task is allowed to access, one can
+//! run the `humility pmp` command, which shows the memory regions that have
+//! been granted u mode access.
+//!
+//! ```console
+//! % humility pmp
+//! humility: attached via OpenOCD
+//! DESC       LOW          HIGH          SIZE ATTR  MODE
+//! pmpaddr00   0x106000 -   0x107fff    2000 r-x-  NAPOT
+//! pmpaddr01   0x141800 -   0x141bff     400 rw--  NAPOT
+//! pmpaddr02        0x0 -       0x1f      20 ----  NAPOT
+//! pmpaddr03        0x0 -       0x1f      20 ----  NAPOT
+//! pmpaddr04        0x0 -       0x1f      20 ----  NAPOT
+//! pmpaddr05        0x0 -       0x1f      20 ----  NAPOT
+//! pmpaddr06        0x0 -       0x1f      20 ----  NAPOT
+//! pmpaddr07        0x0 -       0x1f      20 ----  NAPOT
+//! ```
+//!
+
+use anyhow::{bail, Result};
+use bit_field::BitField;
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use humility::core::Core;
+use humility::hubris::*;
+use humility::regs::rv::RVRegister;
+use humility_cmd::{Archive, Attach, Command, Run, Validate};
+use riscv::register::{Mode, PmpAddr, PmpCfg};
+use std::iter::zip;
+
+#[derive(Parser, Debug)]
+#[clap(name = "pmp", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct PmpArgs {}
+
+fn pmpcmd(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    _subargs: &[String],
+) -> Result<()> {
+    match hubris.arch.as_ref().unwrap().get_e_machine() {
+        goblin::elf::header::EM_RISCV => (),
+        _ => bail!("`humility pmp` only supports riscv"),
+    }
+
+    // place for all the pmpaddr
+    let mut pmpaddrs = Vec::new();
+
+    // read out all the pmpaddr csr using the csr number
+    let base_addr: u32 = RVRegister::PMPADDR0 as u32;
+    let end_addr: u32 = RVRegister::PMPADDR63 as u32;
+    // try to read all of the pmpaddr registers
+    for reg in base_addr..end_addr {
+        let pmpaddr = core.read_reg(
+            // unwrap should always pass since pmpaddr are continuous
+            hubris.arch.as_ref().unwrap().register_from_id(reg).unwrap(),
+        );
+        // not all pmpaddrs will be implemented, so a read may fail
+        // this means the csr is not implemented
+        // so stop after the last valid csr.
+        if let Err(_err) = pmpaddr {
+            break;
+        }
+        let pmpaddr: PmpAddr = (pmpaddr? as usize).into();
+        pmpaddrs.push(pmpaddr);
+    }
+
+    // repeat with pmpcfgs
+    let mut pmpcfgcsrs = Vec::new();
+    let base_addr: u32 = RVRegister::PMPCFG0 as u32;
+    let end_addr: u32 = RVRegister::PMPCFG15 as u32;
+
+    // read all the pmpcfgs
+    for reg in base_addr..end_addr {
+        let csr = core.read_reg(
+            hubris.arch.as_ref().unwrap().register_from_id(reg).unwrap(),
+        );
+        // if the pmpcfg is unavaliable, then we have reached the end of the implemented csrs
```

Comment:
On RV64, only the even-numbered pmpcfg registers are present. Are you planning to rework this later to support RV64 differently?

Reply:
Hm, that's a good point.  I will change it so it checks for 2 unimplemented csrs before bailing.  Then it should be able to skip every other csr on rv64.

Reply:
It will still need some work on rv64, but mainly just calling to some architecture helper to determine bytes per word.

Reply:
Ack